<a href="https://colab.research.google.com/github/samuel-guedes/data_science/blob/master/An%C3%A1lise_Census.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Previsão de Renda - Classificação**

<br>
Este é meu primeiro projeto que estou criando acompanhando o Curso **Machine Learning e Data Science com Python** (Jones Granatyr: disponível em [Udemy](https://https://www.udemy.com/course/machine-learning-e-data-science-com-python-y/)).

A intenção deste primeiro projeto é compartilhar minha evolução em análise de dados partindo do zero e praticar documentação da análise com storytelling. 
Este primeiro projeto trata-se de estudo de predição por Classificação para determinar se uma pessoa ganha mais de 50 mil por ano. A  [base de dados](https://archive.ics.uci.edu/ml/datasets/Adult)  foi baixada do repositório de Machine Learning da [UCI](https://archive.ics.uci.edu/ml/index.php)  (University of California, Irvine).

<br><br>
 
 <br><br>
 <p align=center><img src= "https://bit.ly/3go4sDR" width= "80%"></p>

<br><br>



### **Pré Processamento**

In [ ]:
#Importandos os pacotes necessários

import pandas as pd
import numpy as np

from sklearn.preprocessing import  LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split


In [ ]:
base = pd.read_csv('census.csv')
base.head()

age          workclass  final-weight  ... hour-per-week  native-country  income
0   39          State-gov         77516  ...            40   United-States   <=50K
1   50   Self-emp-not-inc         83311  ...            13   United-States   <=50K
2   38            Private        215646  ...            40   United-States   <=50K
3   53            Private        234721  ...            40   United-States   <=50K
4   28            Private        338409  ...            40            Cuba   <=50K

[5 rows x 15 columns]

In [ ]:
base.dtypes

age                int64
workclass         object
final-weight       int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loos       int64
hour-per-week      int64
native-country    object
income            object
dtype: object

In [ ]:
#Separando as variáveis preditoras e classificação
previsores = base.iloc[:, 0:14].values
classe = base.iloc[:, 14].values
print("Previsores:\n",previsores, "\n\n","Classe:\n",classe)


Previsores:
 [[39 ' State-gov' 77516 ... 0 40 ' United-States']
 [50 ' Self-emp-not-inc' 83311 ... 0 13 ' United-States']
 [38 ' Private' 215646 ... 0 40 ' United-States']
 ...
 [58 ' Private' 151910 ... 0 40 ' United-States']
 [22 ' Private' 201490 ... 0 20 ' United-States']
 [52 ' Self-emp-inc' 287927 ... 0 40 ' United-States']] 

 Classe:
 [' <=50K' ' <=50K' ' <=50K' ... ' <=50K' ' <=50K' ' >50K']


In [ ]:
# transformando variável Nominal em Discreta

column_tranformer = ColumnTransformer([('one_hot_encoder', OneHotEncoder(), [1, 3, 5, 6, 7, 8, 9, 13])],remainder='passthrough')
previsores = column_tranformer.fit_transform(previsores).toarray()


labelencoder_classe = LabelEncoder()
classe = labelencoder_classe.fit_transform(classe)

In [ ]:
print("Previsores:\n",previsores, "\n\n","Classe:\n",classe)

Previsores:
 [[0.0000e+00 0.0000e+00 0.0000e+00 ... 2.1740e+03 0.0000e+00 4.0000e+01]
 [0.0000e+00 0.0000e+00 0.0000e+00 ... 0.0000e+00 0.0000e+00 1.3000e+01]
 [0.0000e+00 0.0000e+00 0.0000e+00 ... 0.0000e+00 0.0000e+00 4.0000e+01]
 ...
 [0.0000e+00 0.0000e+00 0.0000e+00 ... 0.0000e+00 0.0000e+00 4.0000e+01]
 [0.0000e+00 0.0000e+00 0.0000e+00 ... 0.0000e+00 0.0000e+00 2.0000e+01]
 [0.0000e+00 0.0000e+00 0.0000e+00 ... 1.5024e+04 0.0000e+00 4.0000e+01]] 

 Classe:
 [0 0 0 ... 0 0 1]


In [ ]:
#escaler = StandardScaler()
#previsores = escaler.fit_transform(previsores)
#previsores

In [ ]:
#Separando a base de dados entre Treinamento e Teste

previsores_treinamento, previsores_teste, classe_treinamento,classe_teste = train_test_split(previsores, classe, test_size=0.15, random_state=0)
print(previsores_treinamento.shape,previsores_teste.shape, classe_treinamento.shape, classe_teste.shape)

(27676, 108) (4885, 108) (27676,) (4885,)


### **Aprendizagem  de Máquina**

Neste módulo do curso foi aprendido sobre o aprendizado de máquina, que em sua essência é identificar um padrão nos dados históricos entre os dados preditores (atributos) e a Classe. Desta forma, será possível prever uma classificação para um novo registro analisando seus atributos à luz deste padrão identificado.

Os dados de treinamentos são utilzados para que o algoritmo aprenda o padrão entre os atributos e classe. E então com os dados de teste (dados que o algoritmo não conhece), o algoritmo analisa seus atributos e preve classe. Como as classes destes registros são conhecidas, é possível saber qual a precisão da predição - Aprendizado Supervisionado.

Após a previsão é interessante repassar o pré-processamento para avaliar como cada etapa pode afetar a precisão do modelo. Assim, pode-se chegar na melhor performance do modelo baseado no algoritmo.

Neste estudo foram utilizados três algoritmos para a previsão: **Naive Bayes**, **Decision Tree** e **Random Forest**.




#### **Naive Bayes**

Documentação: https://bit.ly/3lA3JDB



In [ ]:
from sklearn.naive_bayes import GaussianNB
classificador = GaussianNB()

#Treinamento:
classificador.fit(previsores_treinamento, classe_treinamento)

#Predição:
previsoes = classificador.predict(previsores_teste)
print(previsoes,classe_teste)

[0 0 0 ... 0 0 0] [0 0 0 ... 0 0 0]


In [ ]:
from sklearn.metrics import  confusion_matrix, accuracy_score
precisao = accuracy_score(classe_teste, previsoes)
matriz = confusion_matrix(classe_teste, previsoes)

print("Precisão: ",precisao, "\n")
print("Matriz Confusão: \n", matriz)


Precisão:  0.7950870010235415 

Matriz Confusão: 
 [[3515  178]
 [ 823  369]]



#### **Árvore de Decisão**

Documentação: https://bit.ly/32ELqUT 


In [ ]:
from sklearn.tree import DecisionTreeClassifier
classificador = DecisionTreeClassifier(criterion='entropy', random_state=0)

#Treinamento:
classificador.fit(previsores_treinamento, classe_treinamento)

#Predição:
previsoes = classificador.predict(previsores_teste)
print(previsoes)

[0 0 0 ... 0 0 1]


In [ ]:
precisao = accuracy_score(classe_teste, previsoes)
matriz = confusion_matrix(classe_teste, previsoes)

print("Precisão: ", precisao,"\n")
print("Matriz Cosnfusão:\n", matriz)

Precisão:  0.8102354145342886 

Matriz Cosnfusão:
 [[3226  467]
 [ 460  732]]


* Random Forest

#### **Random Forest**

Documentação: https://bit.ly/3hW65dP



In [ ]:
from sklearn.ensemble import RandomForestClassifier
classificador = RandomForestClassifier(n_estimators=40, criterion='entropy', random_state=0)

#Treinamento:
classificador.fit(previsores_treinamento, classe_treinamento)

#Predição:
previsoes = classificador.predict(previsores_teste)
previsoes

array([0, 0, 0, ..., 0, 0, 1])

In [ ]:
precisao = accuracy_score(classe_teste, previsoes)
matriz = confusion_matrix(classe_teste, previsoes)

print("Precisão: ", precisao,"\n")
print("Matriz Cosnfusão:\n", matriz)

Precisão:  0.8489252814738997 

Matriz Cosnfusão:
 [[3422  271]
 [ 467  725]]


Comparando a precisão em prever a classificação, o algoritmo Random Forest se mostrou mais eficiente em acertar a classificação entre os três estudados, 84.9% de acerto.

A **Matriz Confusão** nos mostra a relação entre as previsões corretas, falso positivo e falso negativo.

<br><br>
### **Regressão**
**Em andamento*
<br><br>
### **Máquina de Vetores de Suporte (SVM)**
**Em andamento*
<br><br>
### **Redes Neurais Artificiais**
**Em andamento*
<br>